research/picai/fl_nnunet/nnunet_client.py

```diff
@@ -0,0 +1,607 @@
+    """Wraps the nnUNetDataLoader class using the pytorch dataloader to avoid typing errors"""
+
+    def __init__(self, nnunet_dataloader: nnUNetDataLoaderBase, nnunet_config: str, infinite: bool = True) -> None:
+        """
+        Wraps nnunet dataloader classes to make them pytorch compatible
+
+        Args:
+            nnunet_dataloader (nnUNetDataLoaderBase): The nnunet dataloader
+            nnunet_config (str): The nnunet config
+            infinite (bool, optional): Whether or not to treat the dataset
+                as infinite. Defaults to True.
+        """
+        self.nnunet_dataloader = nnunet_dataloader
+
+        # Figure out if dataloader is 2d or 3d
+        self.num_spatial_dims = get_num_spatial_dims(nnunet_config)
+
+        # nnUNetDataloaders store their datasets under the self.data attribute
+        self.dataset: nnUNetDataset = self.nnunet_dataloader.generator._data
+        super().__init__(dataset=self.dataset, batch_size=self.nnunet_dataloader.generator.batch_size)
+
+        # nnunet dataloaders are infinite by default so we have to track steps to stop iteration
+        self.current_step = 0
+        self.infinite = infinite
+        self.len = len(self.dataset)
+
+    def __next__(self) -> Tuple[torch.Tensor, Union[torch.Tensor, Dict[str, torch.Tensor]]]:
+        if not self.infinite and self.current_step == self.len:
+            self.reset()
+            raise StopIteration
+        else:
+            self.current_step += 1
+            batch = next(self.nnunet_dataloader)  # This returns a dictionary
+            # Note: When deep supervision is on, target is a list of segmentations at various scales
+            # nnUNet has a wrapper for loss functions to enable deep supervision
+            inputs: torch.Tensor = batch["data"]
+            if isinstance(batch["target"], list):
+                target_dict = convert_ds_list_to_dict(batch["target"], self.num_spatial_dims)
+                return inputs, target_dict
+            elif isinstance(batch["target"], torch.Tensor):
+                target_tensor: torch.Tensor = batch["target"]
+                return inputs, target_tensor
+            else:
+                raise TypeError(
+                    "Was expecting the target generated by the nnunet dataloader to be a list or a torch.Tensor"
+                )
+
+    def __len__(self) -> int:
+        # Should return num_samples // batch_size
+        # nnUNetDataloaders are 'infinite' meaning they randomly sample batches
+        # from the dataset. This makes the distinction between
+        # num_samples // batch_size and ceil(num_samples/batch_size)
+        # meaningless. We will abritrarily use the later
+        num_samples = len(self.dataset)
+        batch_size = self.nnunet_dataloader.generator.batch_size
+        return int(ceil(num_samples / batch_size))
+
+    def reset(self) -> None:
+        self.current_step = 0
+
+    def __iter__(self) -> DataLoader:  # type: ignore
```

Comment:
This does not resolve the issue. The parent class expects the return type to be_BaseDataLoader.

Reply:
I see. I looked into the internals here a bit more. Is there a reason we want to override the standard `__iter__` functionality of the parent class here? It looks like the `__iter__` function in the DataLoader class is somewhat intricate and doesn't return the dataloader itself as part of the process. I removed this override and mypy doesn't have any issues, but perhaps there is a deeper reason why you went this route.

Reply:
I can't actually recall why the method is overridden. I'd me more than happy to remove it, I'll test and confirm that it is not needed

Reply:
Ok so turns out that we need to override the __iter__ function to get the dataloader to work. The main reason being that I believe the parent method assumes the dataset has some key or attribute that it doesn't. That is because the self.dataset attribute is type nnUNetDataset which does not inherit from any other class. The pytorch dataloader's __iter__ method return an object that inherits most of its functionality from _BaseDataLoaderIter. This class is is in turn a wrapper on the pytorch Dataloader class. During initialization of _BaseDataLoaderIter is where we get errors. Since we are defining our own next method, we don't need all this complexity (The pytorch Dataloader class does not have a __next__ method, this is all defined by the _BaseDataLoaderIter class).

Reply:
Actually the reason it is not working might be because we are overriding the next method. I think pytorch just wants to have control of all that and just wants you to define a dataset class with a get_item method. Because we don't want to reimplement all the data_preprocessing and formatting that occurs in the nnunet_dataloaders default next method i think this is the easiest solution. There might be a way to reformat all this code somehow to get it to play nicer with pytorch's dataloading but i would have to look into it